### PR TITLE
ORC-372: Enable valgrind for C++ travis-ci tests

### PR DESCRIPTION
**Pull request by majetideepak**

### CODE REVIEW

**Comment by majetideepak** [comment]
There are indeed valgrind failures. I will push a followup patch to fix these.

**Comment by majetideepak** [comment]
@wgtmac, @xndai  can you take a look at this patch? ZLIB compression code seems to be leaking memory. Thanks!

File: c++/src/TypeImpl.cc [hunk 1]
```diff
@@ -258,31 +258,34 @@ namespace orc {
 
     case STRUCT: {
       StructVectorBatch *result = new StructVectorBatch(capacity, memoryPool);
+      std::unique_ptr<ColumnVectorBatch> return_value = std::unique_ptr<ColumnVectorBatch>(result);
```
Comment (wgtmac) [inline c++/src/TypeImpl.cc]: maybe we can directly merge line 260 and 261 to a single statement. same for below.

File: c++/src/TypeImpl.cc [hunk 2]
```diff
@@ -498,54 +506,54 @@ namespace orc {
...
+    std::vector<std::pair<std::string, ORC_UNIQUE_PTR<Type> > > res =
       TypeImpl::parseType(input, 0, input.size());
     if (res.size() != 1) {
       throw std::logic_error("Invalid type string.");
     }
-    return ORC_UNIQUE_PTR<Type>(res[0].second);
+    return std::move(res[0].second);
   }
 
   Type* TypeImpl::parseArrayType(const std::string &input,
```
Comment (wgtmac) [inline c++/src/TypeImpl.cc]: is it better to change return type to unique_ptr as well to be consistent? same for below.

File: c++/src/Compression.cc [hunk 3]
```diff
@@ -282,6 +285,11 @@ DIAGNOSTIC_PUSH
     }
   }
 
+  void ZlibCompressionStream::end() {
+    (void)deflateEnd(&strm);
+  }
+
+
```
Comment (wgtmac) [inline c++/src/Compression.cc]: remove this line

File: c++/src/TypeImpl.cc (hunk 1 above)
Comment (majetideepak) [reply c++/src/TypeImpl.cc]: we need `result` of type `StructVectorBatch` to access fields on line 263

File: c++/src/Compression.cc [hunk 4]
```diff
@@ -199,13 +199,16 @@ namespace orc {
                           uint64_t blockSize,
                           MemoryPool& pool);
 
+    ~ZlibCompressionStream() { end(); }
```
Comment (omalley) [inline c++/src/Compression.cc]: You need an override on this or the mac complains.

File: c++/test/CMakeLists.txt [hunk 5]
```diff
@@ -64,7 +64,12 @@ target_link_libraries (create-test-files
   protobuf
 )
 
-add_test (NAME orc-test COMMAND orc-test)
+if (TEST_VALGRIND_MEMCHECK)
+  add_test (orc-test
```
Comment (omalley) [inline c++/test/CMakeLists.txt]: You have trailing spaces here.

File: c++/src/Compression.cc (hunk 4 above)
Comment (majetideepak) [reply c++/src/Compression.cc]: Thanks!

**Pull request closed without merging**